servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,624 @@
+                    return false;
+                } else if (pendingDemandUpdater.compareAndSet(this, prevDemand, prevDemand - 1)) {
+                    return true;
+                }
+            }
+        }
+
+        private void tryEmitItem(Object item, FlatMapPublisherSubscriber<T, R> subscriber) {
+            if (tryAcquireLock(emittingLockUpdater, this)) { // fast path. no concurrency, try to skip the queue.
+                boolean mappedSourcesCompleted = false;
+                try {
+                    if (subscriber.hasSignalsQueued() || (needsDemand(item) && !tryDecrementPendingDemand())) {
+                        subscriber.markSignalsQueued();
+                        enqueueItem(item); // drain isn't necessary. when demand arrives a drain will be attempted.
+                    } else if (item == MAPPED_SOURCE_COMPLETE) {
+                        mappedSourcesCompleted = true;
+                    } else {
+                        final boolean demandConsumed = sendToTarget(item);
+                        assert demandConsumed == needsDemand(item);
+                    }
+                } finally {
+                    if (!releaseLock(emittingLockUpdater, this)) {
+                        drainPending();
+                    }
+                }
+                if (mappedSourcesCompleted) { // request more from upstream outside the critical section.
+                    requestMoreFromUpstream(1);
+                }
+            } else { // slow path. there is concurrency, go through the queue to avoid concurrent delivery.
+                subscriber.markSignalsQueued();
+                enqueueAndDrain(item);
+            }
+        }
+
+        private void enqueueItem(Object item) {
+            if (!signals.offer(item)) {
+                enqueueFailed(item);
+            }
+        }
+
+        private void enqueueAndDrain(Object item) {
+            enqueueItem(item);
+            drainPending();
+        }
+
+        private void enqueueFailed(Object item) {
+            QueueFullException exception = new QueueFullException("signals");
+            if (item instanceof TerminalNotification) {
+                LOGGER.error("Queue should be unbounded, but an offer failed!", exception);
+                throw exception;
+            } else {
+                onError0(exception, false);
+            }
+        }
+
+        private void drainPending() {
+            Throwable delayedCause = null;
+            boolean tryAcquire = true;
+            int mappedSourcesCompleted = 0;
+            while (tryAcquire && tryAcquireLock(emittingLockUpdater, this)) {
+                try {
+                    final long prevDemand = pendingDemandUpdater.getAndSet(this, 0);
+                    long emittedCount = 0;
+                    if (prevDemand < 0) {
+                        pendingDemand = prevDemand;
+                        // Don't wait for demand to deliver the terminalNotification if present. The queued signals
```

Comment:
I don't think this is required, i.e. have dual mode whether errors skip the queue or not based on `pendingDemand` being negative.

The context is spread out to call-sites of `doCancel()` and makes the control flow too hard to follow.

Reply:
discussed details offline and clarified the failure scenarios we are accommodating for with this code. fine as is.